core/state_transition.go

```diff
@@ -413,10 +340,7 @@ func (st *StateTransition) buyGas() error {
 
 	// SoulGasToken doesn't support burning balance of zero account, so here we ensure that `gasFromSoul` is false for zero account.
 	if st.msg.From != (common.Address{}) && st.evm.ChainConfig().IsOptimism() && st.evm.ChainConfig().Optimism.UseSoulGasToken {
-		have, err := st.GetSoulBalance(st.msg.From)
-		if err != nil {
-			return fmt.Errorf("GetSoulBalance error:%v", err)
-		}
+		have := st.GetSoulBalance(st.msg.From)
 		if have, want := have.ToBig(), new(big.Int).Sub(balanceCheck, st.msg.Value); have.Cmp(want) >= 0 {
 			if have, want := st.state.GetBalance(st.msg.From).ToBig(), st.msg.Value; have.Cmp(want) < 0 {
```

Comment:
Does this mean that even if the sender has enough soul gas, it will still return an error if there isn’t enough native gas token?

Reply:
No, that issue is to be patched by modifying `ValidateTransactionWithState` to also consider SGT balance. I'll create a separate PR after merging this one.

Reply:
In what situation that the msg is from 0x0 address?

Reply:
When called from [here](https://github.com/ethereum-optimism/op-geth/blob/f21b97f1d6f673b608f72e9e60d63e70bd925e85/internal/ethapi/api.go#L1340) or https://github.com/ethereum-optimism/op-geth/blob/f21b97f1d6f673b608f72e9e60d63e70bd925e85/internal/ethapi/api.go#L1462. Basically those APIs that doesn't require signature.